compiler/rustc_trait_selection/src/traits/project.rs

```diff
@@ -672,7 +672,10 @@ impl<'a, 'b, 'tcx> TypeFolder<TyCtxt<'tcx>> for AssocTypeNormalizer<'a, 'b, 'tcx
     #[instrument(skip(self), level = "debug")]
     fn fold_const(&mut self, constant: ty::Const<'tcx>) -> ty::Const<'tcx> {
         let tcx = self.selcx.tcx();
-        if tcx.lazy_normalization() || !needs_normalization(&constant, self.param_env.reveal()) {
+        if tcx.features().generic_const_exprs
+            || tcx.trait_solver_next()
```

Comment:
I don't know if it makes sense to do this- we dont stop `fold_ty` from normalizing things if `trait_solver_next` is enabled so I wouldn't think it makes sense to avoid normalization of consts.

Reply:
It doesn't hurt, though? Like, the only reason we have eager normalization of types currently is because it's difficult to disable because of alias-relate ambiguities (like due to lifetimes primarily). 

I could revert this, but it does seem to move us closer to actually-lazy-norm at least for consts?

Reply:
I guess i'm a bit nervous about having types/consts diverge especially when it comes to implementing lazy norm since if we only do it for consts it seems more likely that bugs would go hidden since usage of const generics is way rarer than usage of type generics.

Reply:
I think i'd want @lcnr to be okay with doing this if we are going to

Reply:
That's fair, I can revert it and we can discuss "if and how we want to get to actually-lazy-norm-for-consts" in the new solver separately then.